Clean up usage of ArrayPool in the Cryptography assemblies

bartonjs opened the pull request:
Moving cryptography to its own ArrayPool reduces both the chances of
pool mis-use within cryptography to share secrets and pool-mis-use outside of
cryptography from making an area that's already hard enough to debug
problems in even harder to debug.

Unlike ArrayPool, CryptoPool's default return is clear-everything, but provides
an accelerator pattern for selective-clear-and-return.

Places where the arrays get passed to unsealed virtuals continue to use the
shared ArrayPool.

CryptoStream was also changed to not return the pool arrays when
exceptions are thrown, because it's hard to reason about the lifetime with
exceptions, unsealed virtuals, and Task state.

krwq commented on src/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.PooledCrypto.cs:
```diff
@@ -0,0 +1,40 @@
+
+            if (ret != 1)
+            {
+                CryptoPool.Return(bytes, clearSize: 0);
```
why not clear in case of error?

bartonjs replied:
I guess it could.  I was just maintaining parity (which was more obvious before I moved this to a new file to avoid an "included in too many assemblies" problem).

GrabYourPitchforks commented:
Turns out that unlike the shared pool, `ArrayPool<T>.Create(...)`-instantiated pools aren't culled by the GC at regular intervals. In theory this means that an array in a static pool could survive for the lifetime of the application, which could potentially result in memory leaks if an attacker can influence the size of the rental. I don't think we have any code patterns throughout the crypto stack which would be affected by this, but it's good to keep in mind.

GrabYourPitchforks commented on src/Common/src/System/Security/Cryptography/CryptoPool.cs:
```diff
@@ -0,0 +1,31 @@
+
+            if (!clearWholeArray && clearSize != 0)
+            {
+                Array.Clear(array, 0, clearSize);
```
`CryptographicOperations.ZeroMemory` if available?

bartonjs replied:
It's not available in all of the configurations the file gets compiled into, and didn't seem worth the \#if.  Even if we have the "eliminate writes without reads" optimization that shouldn't be able to kick in since the array is then passed to pool.Return.

GrabYourPitchforks commented:
Now I'm second-guessing my statement above re: attackers not being able to influence the size of the rental. Do components like the OpenSSL transforms allow this?

jkotas commented:
>  In theory this means that an array in a static pool could survive for the lifetime of the application,

Right, each instance of the array can hold up to 50MB+ of memory that is not going away. If I am counting right, you are adding 8 separate instances of these - so this is a potential 400MB+ memory leak.

If you want to do this, I think there needs to be a performance work done on the custom pools first.

bartonjs commented:
>  so this is a potential 400MB+ memory leak

N easy reductions:

* Reflect-instantiate the internal type that drives ArrayPool.Shared
* Reflect-find the pool from Algorithms so they're all the same pool
* Change the pool static to a WeakReference

Any favorites?

jkotas commented:
> Reflect-instantiate the internal type that drives ArrayPool.Shared

That pool is not designed and tested to be have multi-instances. If it does not crash, you will actually end up using the shared pool because of statics.

> Reflect-find the pool from Algorithms so they're all the same pool

That does not work well with IL linker.

> Change the pool static to a WeakReference

This should work ok.

stephentoub commented on src/Common/src/System/Security/Cryptography/CryptoPool.cs:
```diff
@@ -0,0 +1,31 @@
+    {
+        internal const int ClearAll = -1;
+
+        private static readonly ArrayPool<byte> s_pool = ArrayPool<byte>.Create();
```
This isn't just introducing one separate pool... it's introducing one separate pool per assembly that imports this file, which looks to be ~10.

I question whether this change is really worthwhile.  I read your justification, but I don't see why these libraries are significantly more at risk than every other library we have that uses the shared pool.  Everything is in process, where there's no security boundary.  We're already being a little paranoid and aggressively clearing when we put the arrays back.  As Levi pointed out, the pool created by Create is also less sophisticated and doesn't trim.  It also has a difference performance profile, so any measurements we did in the past around throughput might be impacted.

stephentoub commented:
> That pool is not designed and tested to be have multi-instances. If it does not crash, you will actually end up using the shared pool because of statics.

In particular, it uses [ThreadLocal]s, so to Jan's point, you'll actually end up using some of the same state.

>> Change the pool static to a WeakReference
> This should work ok (it may end up to be worse than just not using pool at all in some cases).

I imagine in many cases it'll be worse than not using the pool in most cases.  Seems like the pool would end up getting collected during most GCs, at which point we not only incur most of the allocations we would if we didn't use the pool, but we artifically promote more arrays to being in higher generations, and we incur the overheads of the pool itself.

bartonjs commented:
Since I hadn't come up with a solution to the infinite-array-lifetime problem (well, a few in my head, but none that made it through fingers and onto a keyboard), the PR is now offering the following benefits:

* It makes a common clear-and-return pattern in the crypto libraries.
  * FWIW, this clearing isn't "paranoia", it's part of a customer need for certain customers (which, without it, stop using .NET).
* Moves a couple of places to use opportunistic stackalloc instead of (or before) renting.
* Changes a nontrivial rent-write-call-copy to use a pattern like string.Create to avoid renting (and copying) altogether.
* Fixes places that rented to a local then promoted to a field to rent to the field to avoid exceptional double-return.
* Added a ValueEquals to AsnWriter to avoid a drain-to-compare (and the renting associated with that)
* Fixed (at least) one rent-with-no-return to return.
* Removed array allocations in Rfc2898DeriveBytes that could have just been on the existing carryover buffer field.
* Moves (at least) one async rent try/finally-return to only return on success to avoid composed-with-an-aborted-Task use-after-return.
* And, of course, leaves things set up to facilitate switching crypto internals to a private pool at a later date.

I can slice out some of the changes to separate items, if preferred.  If there's consensus for OK-as-is then I'll just retitle and redescribe it as cleaning up ArrayPool usage within the crypto libraries.

stephentoub commented on src/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.PooledCrypto.cs:
```diff
@@ -0,0 +1,40 @@
+            return RentDynamicBuffer((ptr, buf, i) => GetAsn1StringBytes(ptr, buf, i), asn1);
+        }
+
+        private static ArraySegment<byte> RentDynamicBuffer<THandle>(NegativeSizeReadMethod<THandle> method, THandle handle)
```
Is anything else using this?  Wondering if it actually needs to be another generic method that operates on a delegate or whether its implementation could just be part of RentAsn1StringBytes.

bartonjs replied:
Nothing else is using it currently.  But no analysis was done to see what things using the allocating version would benefit from the renting version; so it seems like leaving it like this would save the effort later.

stephentoub commented on src/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.Encode.cs:
```diff
@@ -69,7 +70,7 @@ internal static partial class Crypto
                 // Since we don't know what was written, assume it was secret and clear the value.
                 // (It doesn't matter much, since we're behind Debug.Fail)
-                ArrayPool<byte>.Shared.Return(data, clearArray: true);
+                CryptoPool.Return(data, CryptoPool.ClearAll);
```
ClearAll is the default, right?  Does this need to be specified?

stephentoub commented on src/Common/src/System/Security/Cryptography/Asn1Reader/AsnReader.Integer.cs:
```diff
@@ -107,9 +107,7 @@ public BigInteger ReadInteger(Asn1Tag expectedTag)
-                // Clear the whole tmp so that not even the sign bit is returned to the array pool.
-                Array.Clear(tmp, 0, tmp.Length);
-                ArrayPool<byte>.Shared.Return(tmp);
+                CryptoPool.Return(tmp, CryptoPool.ClearAll);
```
Same ClearAll question.  Seems like either we should be explicit at all call sites, in which case we should make the parameter non-optional, or we should remove the argument when it's the default.

stephentoub commented on src/Common/src/System/Security/Cryptography/Asn1Reader/AsnReader.Integer.cs:
```diff
@@ -107,9 +107,7 @@ public BigInteger ReadInteger(Asn1Tag expectedTag)
             }
             finally
             {
-                // Clear the whole tmp so that not even the sign bit is returned to the array pool.
```
Presumably the comment was here to prevent some future dev from "optimizing" this?  Is it worth keeping for that reason?

stephentoub commented on src/Common/src/System/Security/Cryptography/Asn1Reader/AsnWriter.BitString.cs:
```diff
@@ -120,20 +117,190 @@ private void WriteBitStringCore(Asn1Tag tag, ReadOnlySpan<byte> bitString, int u
 
-        // T-REC-X.690-201508 sec 9.2, 8.6
-        private void WriteConstructedCerBitString(Asn1Tag tag, ReadOnlySpan<byte> payload, int unusedBitCount)
+#if netcoreapp || uap || NETCOREAPP
```
Why do we have two different versions of netcoreapp vs NETCOREAPP?  Can we fix that to just have one compilation constant for it?

bartonjs replied:
One of them is from corefx, the other is corefxlab, from when the code gets copied back and forth as part of the Asn1.Experimental package.

stephentoub commented on src/Common/src/System/Security/Cryptography/Asn1Reader/AsnWriter.BitString.cs:
```diff
@@ -120,20 +117,190 @@ private void WriteBitStringCore(Asn1Tag tag, ReadOnlySpan<byte> bitString, int u
+                _offset = savedOffset;
+                scratchSpace.Clear();
+
+                // TODO: Probably warrants a distinct message.
```
Are these TODOs tracked somewhere?

bartonjs replied:
Just with their existence :smile:.  It's mainly notes for when the types become public, and should get a bit friendlier.

stephentoub commented on src/Common/src/System/Security/Cryptography/Asn1Reader/AsnWriter.cs:
```diff
@@ -57,7 +57,8 @@ public void Dispose()
             {
                 Array.Clear(_buffer, 0, _offset);
 #if !CHECK_ACCURATE_ENSURE
```
In what situation is this constant set?  Are we leaking an array in that case?  Could we just always return the array via a call to `CryptoPool.Return(_buffer, _offset)` and remove the Clear call?

bartonjs replied:
It's just manually set, it's a testing tool to make sure that writing after calling EnsureWriteCapacity is not accidentally exceeding bounds, by making it not give more than was requested.

https://github.com/dotnet/corefx/blob/d8797a0f1616c1db1e5294964f2826e051f2615e/src/Common/src/System/Security/Cryptography/Asn1Reader/AsnWriter.cs#L214-L224

stephentoub commented on src/Common/src/System/Security/Cryptography/CngPkcs8.cs:
```diff
@@ -250,8 +250,8 @@ internal static unsafe Pkcs8Response ImportPkcs8PrivateKey(ReadOnlySpan<byte> so
-                    ArrayPool<byte>.Shared.Return(rented);
-                    rented = ArrayPool<byte>.Shared.Rent(checked(size * 2));
+                    CryptoPool.Return(rented, rentWritten);
+                    rented = CryptoPool.Rent(checked(size * 2));
```
If Rent fails (e.g. OOM), we could end up returning `rented` twice, I think, once here and once in the finally.

stephentoub commented on src/Common/src/System/Security/Cryptography/CngPkcs8.cs:
```diff
@@ -287,8 +286,8 @@ internal static unsafe Pkcs8Response ImportPkcs8PrivateKey(ReadOnlySpan<byte> so
-                    ArrayPool<byte>.Shared.Return(rented);
-                    rented = ArrayPool<byte>.Shared.Rent(checked(size * 2));
+                    CryptoPool.Return(rented, rentWritten);
+                    rented = CryptoPool.Rent(checked(size * 2));
```
Ditto

stephentoub commented on src/Common/src/System/Security/Cryptography/DSACng.SignVerify.cs:
```diff
@@ -81,29 +68,22 @@ public override bool VerifySignature(byte[] rgbHash, byte[] rgbSignature)
-                    }
-                }
-                finally
+                Span<byte> stackBuf = stackalloc byte[WindowsMaxQSize];
```
Nice to see these pool uses eliminated via stack space.

stephentoub commented on src/Common/src/System/Security/Cryptography/PasswordBasedEncryption.cs:
```diff
@@ -418,8 +415,7 @@ private static CryptographicException AlgorithmKdfRequiresChars(string algId)
                         {
-                            CryptographicOperations.ZeroMemory(sourceRent.AsSpan(0, source.Length));
-                            ArrayPool.Return(sourceRent);
+                            CryptoPool.Return(sourceRent, source.Length);
```
The stronger version of zeroing isn't needed anymore?  I'm fine with that, just confirming it's intentional.

bartonjs replied:
Based on having received feedback on this twice, I've gone ahead and done the \#if in CryptoPool to use CryptographicOperations.ZeroMemory when it's available.

stephentoub commented on src/Common/src/System/Security/Cryptography/RsaPaddingProcessor.cs:
```diff
@@ -457,8 +456,8 @@ internal bool VerifyPss(ReadOnlySpan<byte> mHash, ReadOnlySpan<byte> em, int key
-                dbMask.Clear();
-                ArrayPool<byte>.Shared.Return(dbMaskRented);
+                CryptographicOperations.ZeroMemory(dbMask);
+                CryptoPool.Return(dbMaskRented);
```
Should this have `clearSize: 0`?

stephentoub commented on src/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/DSA.cs:
```diff
@@ -199,23 +201,34 @@ public virtual bool VerifyData(ReadOnlySpan<byte> data, ReadOnlySpan<byte> signa
             {
-                int hashLength = 0;
-                byte[] hash = ArrayPool<byte>.Shared.Rent(i);
+                // Use ArrayPool.Shared because the array is passed out.
```
Nit: maybe clarify to say "Use ArrayPool.Shared instead of CryptoPool"
